src/transformers/models/superpoint/modeling_superpoint.py

```diff
@@ -0,0 +1,577 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""PyTorch SuperPoint model."""
+
+from typing import Optional, Tuple, Union
+
+import torch
+from torch import nn
+
+from transformers import PreTrainedModel
+from transformers.modeling_outputs import (
+    BaseModelOutputWithNoAttention,
+    ImagePointDescriptionOutput,
+)
+from transformers.models.superpoint.configuration_superpoint import SuperPointConfig
+
+from ...utils import (
+    add_code_sample_docstrings,
+    add_start_docstrings,
+    add_start_docstrings_to_model_forward,
+    logging,
+    replace_return_docstrings,
+)
+
+
+logger = logging.get_logger(__name__)
+
+_CONFIG_FOR_DOC = "SuperPointConfig"
+
+_CHECKPOINT_FOR_DOC = "stevenbucaille/superpoint"
+
+
+SUPERPOINT_PRETRAINED_MODEL_ARCHIVE_LIST = ["stevenbucaille/superpoint"]
+
+
+class SuperPointEncoder(nn.Module):
+    """
+    SuperPoint encoder module. It is made of 4 convolutional layers with ReLU activation and max pooling, reducing the
+     dimensionality of the image.
+    """
+
+    def __init__(self, config: SuperPointConfig) -> None:
+        super().__init__()
+        self.hidden_sizes = config.hidden_sizes
+        self.descriptor_dim = config.descriptor_dim
+
+        self.relu = nn.ReLU(inplace=True)
+        self.pool = nn.MaxPool2d(kernel_size=2, stride=2)
+
+        self.conv1a = nn.Conv2d(1, self.hidden_sizes[0], kernel_size=3, stride=1, padding=1)
+        self.conv1b = nn.Conv2d(
+            self.hidden_sizes[0],
+            self.hidden_sizes[0],
+            kernel_size=3,
+            stride=1,
+            padding=1,
+        )
+        self.conv2a = nn.Conv2d(
+            self.hidden_sizes[0],
+            self.hidden_sizes[1],
+            kernel_size=3,
+            stride=1,
+            padding=1,
+        )
+        self.conv2b = nn.Conv2d(
+            self.hidden_sizes[1],
+            self.hidden_sizes[1],
+            kernel_size=3,
+            stride=1,
+            padding=1,
+        )
+        self.conv3a = nn.Conv2d(
+            self.hidden_sizes[1],
+            self.hidden_sizes[2],
+            kernel_size=3,
+            stride=1,
+            padding=1,
+        )
+        self.conv3b = nn.Conv2d(
+            self.hidden_sizes[2],
+            self.hidden_sizes[2],
+            kernel_size=3,
+            stride=1,
+            padding=1,
+        )
+        self.conv4a = nn.Conv2d(
+            self.hidden_sizes[2],
+            self.hidden_sizes[3],
+            kernel_size=3,
+            stride=1,
+            padding=1,
+        )
+        self.conv4b = nn.Conv2d(
+            self.hidden_sizes[3],
+            self.hidden_sizes[3],
+            kernel_size=3,
+            stride=1,
+            padding=1,
+        )
```

Comment:
This is too hard coded - it should be possible to have a different number of conv layers here

Rather than structure like this - I would have another class e.g. `ConvBlock` which you can use to remove this repeated code both here an in the forward pass e.g. 

```
class SuperPointConvBlock(nn.Module):
    def __init__(self, config, input_dim, output_dim, add_pooling_layer=False): 
        self.conv_a = nn.Conv2d(input_dim, output_dim, kernel_size=3, stride=1, padding=1)
        self.conv_b = nn.Conv2d(output_dim, output_dim, kernel_size=3, stride=1, padding=1)
        self.relu = nn.Relu()
        self.pooler = nn.MaxPool2d(kernel_size=2, stride=2) if add_pooling_layer else None

    def forward(self, hidden_states):
        x = hidden_states
        x = self.relu(self.conv_a)
        x = self.relu(self.conv_b)

        if self.pooler is not None:
            x = self.pooler(x)

        return x

class SuperPointEncoder(nn.Module):
    def __init__(self, config):
        input_dim = 1

        conv_blocks = []
        for hidden_dim in config.hidden_dims[:-1]: 
            conv_blocks.append(SuperPointConvBlock(input_dim, hidden_dim), add_pooling_layer=True)
        conv_blocks.append(SuperPointConvBlock(input_dim, config.hidden_dims[-1], add_pooling_layer=False))
        self.conv_blocks = nn.ModuleList(conv_blocks)
```

Reply:
Fixed in [this commit](https://github.com/huggingface/transformers/pull/25786/commits/c43052de41132765b2bdfccb863a0ee7aa33a9c4)